peps/pep-0787.rst

```diff
@@ -0,0 +1,279 @@
+
+
+Renderer for shell escaping added to :mod:`shlex`
+-------------------------------------------------
+
+As a reference implementation, a renderer for safe POSIX shell escaping will be added to
+the :mod:`shlex` module. This renderer would be called ``sh`` and would be equivalent to
+calling ``shlex.quote`` on each field value in the template literal.
+
+Thus:
+
+.. code-block:: python
+
+  os.system(shlex.sh(t"cat {myfile}"))
+
+would have the same behavior as:
+
+.. code-block:: python
+
+  os.system("cat " + shlex.quote(myfile)))
+
+The implementation would be:
+
+.. code-block:: python
+
+  def sh(template: TemplateLiteral):
+      def render_field(value, format_spec, conversion_spec)
+          field_text = format(value, format_spec, conversion_spec)
+          return quote(field_text)
+      return template.render(render_field=render_field)
+
+The addition of ``shlex.sh`` will NOT change the existing admonishments in the
+:mod:`subprocess` documentation that passing ``shell=True`` is best avoided, nor the
+reference from the :func:`os.system` documentation the higher level ``subprocess`` APIs.
+
+Then implementation would look like:
+
+.. code-block:: python
+
+    from string.templatelib import Template, Interpolation
+
+    def sh(template: Template) -> str:
+        parts: list[str] = []
+        for item in template:
+        if isinstance(item, Interpolation):
+            parts.append(quote(str(item.value)))
+        else:
+            parts.append(item)
+
+        return "".join(parts)
+
+This allows for explicit escaping of t-strings for shell usage:
+
+.. code-block:: python
+
+    import shlex
+    # Safe shell command construction
+    command = shlex.sh(t"cat {filename}")
+    os.system(command)
+
+Changes to subprocess module
+----------------------------
+
+With the additional renderer in the shlex module, and the addition of template strings,
+the :mod:`subprocess` module can be changed to handle accepting template strings
+as an additional input type to ``Popen``, as it already accepts a sequence, or a string,
+with different behavior for each. In return, all :class:`subprocess.Popen` higher level
+functions such as :func:`subprocess.run` could accept strings in a safe way
+(at least on :ref:`POSIX systems <pep-9999-defer-non-posix-shells>`).
+
+For example:
+
+.. code-block:: python
+
+  subprocess.run(t"cat {myfile}", shell=True)
+
+would automatically use the ``shlex.sh`` renderer provided in this PEP. Therefore, using
+``shlex`` inside a ``subprocess.run`` call like so:
+
+.. code-block:: python
+
+  subprocess.run(shlex.sh(t"cat {myfile}"), shell=True)
+
+would be redundant, as ``run`` would automatically render any template literals
+through ``shlex.sh``
+
+Alternatively, when ``subprocess.Popen`` is run without ``shell=True``, it could still
+provide subprocess with a more ergonomic syntax. For example:
+
+.. code-block:: python
+
+  subprocess.run(t"cat {myfile} --flag {value}")
+
+would be equivalent to:
```

Comment:
except it would still invoke a shell?

Reply:
No. Not in this case. Do I need to make that more clear?

Reply:
I am a little confused by the two different implementations for `def sh` above, at least one of which returns a string.

Reply:
But yes it's clear this wouldn't invoke a subshell; you say below that it'd `shlex.split` the string after. We can resolve this.